Assignments/ip2/ip2.md

```diff
@@ -0,0 +1,310 @@
+The base class tracks the game model (`this._model`), the set of players in the game (`this.players`), and the set of observers in the game (`this.observers`). It also provides methods for joining and leaving the game, as well as a base implementation of `_updateFrom`, which is responsible for updating the game state when the backend notifies the frontend of a change.
+
+Your first task is to implement each of the properties and methods of the `TicTacToeAreaController` class (`frontend/src/classes/interactable/TicTacToeAreaController.ts`). The specification for these properties and methods appears below:
+
+{::options parse_block_html="true" /}
+<details><summary markdown="span">View the specification for these methods</summary>
+{% highlight typescript %}
+  {
+  /**
+   * Returns the current state of the board.
+   *
+   * The board is a 3x3 array of TicTacToeCell, which is either 'X', 'O', or undefined.
+   *
+   * The 2-dimensional array is indexed by row and then column, so board[0][0] is the top-left cell,
+   * and board[2][2] is the bottom-right cell
+   */
+  get board(): TicTacToeCell[][]
+  /**
+   * Returns the player with the 'X' game piece, if there is one
+   */
+  get x(): PlayerController | undefined
+  /**
+   * Returns the player with the 'O' game piece, if there is one
+   */
+  get o(): PlayerController | undefined 
+  /**
+   * Returns the number of moves that have been made in the game
+   */
+  get moveCount(): number
+  /**
+   * Returns the winner of the game, if there is one
+   */
+  get winner(): PlayerController | undefined
+  /**
+   * Returns the player whose turn it is, if the game is in progress
+   * Returns undefined if the game is not in progress
+   */
+  get whoseTurn(): PlayerController | undefined
+  /**
+   * Returns true if it is our turn to make a move in the game
+   * Returns false if it is not our turn, or if the game is not in progress
+   */
+  get isOurTurn(): boolean
+  /**
+   * Returns true if the current player is a player in this game
+   */
+  get isPlayer(): boolean
+  /**
+   * Returns the game piece of the current player, if the current player is a player in this game
+   *
+   * Throws an error PLAYER_NOT_IN_GAME_ERROR if the current player is not a player in this game
+   */
+  get gamePiece(): 'X' | 'O'
+  /**
+   * Returns the status of the game.
+   */
+  get status(): GameStatus
+  /**
+   * Returns true if the game is in progress
+   */
+  public isActive(): boolean
+  /**
+   * Updates the internal state of this TicTacToeAreaController to match the new model.
+   *
+   * Calls super._updateFrom, which updates the occupants of this game area and
+   * other common properties (including this._model).
+   *
+   * If the board has changed, emits a 'boardChanged' event with the new board.
+   *
+   * If the turn has changed, emits a 'turnChanged' event with true if it is our turn, and false otherwise.
+   */
+  protected _updateFrom(newModel: GameArea<TicTacToeGameState>): void
```

Comment:
From my TA:

> From the specifications, I assumed that a `boardChanged` event should be emitted only if
> the new board is different from the old board and a `turnChanged` event should be
> emitted only when it was previously x’s turn and now it’s o’s turn and vise versa.
> 
> I see that the implementation and the tests expect the events to be emitted regardless
> and as long as `newModel.game` exists.
> 
> Is it necessary to check for a change in state or can the event be emitted regardless?

Reply:
Yeah, this is a good find - I was lazy :P I updated the spec to clarify that shouldn't be emitted if no change, added tests to check for that, and we'll now be expecting students to check for it in their tests, too.